src/Avalonia.Controls/Control.cs

```diff
@@ -312,13 +313,18 @@ internal void ScheduleOnLoadedCore()
         /// </summary>
         internal void OnLoadedCore()
         {
-            if (_isLoaded == false &&
+            if (_loadState == LoadState.LoadPending &&
                 ((ILogical)this).IsAttachedToLogicalTree)
```

Comment:
Note: changing this to `IsAttachedToVisualTree` also solves the problem (at least the one which this PR tests). `OnLoadedCore/OnUnloadedCore` are always called from `VisualTreeAttached/VisualTreeDetached` so I'm not sure why we're using the logical tree here.

However it's been explicitly changed in #8738 so I didn't touch it. Not raising the event inappropriately should be sufficient.

Reply:
Hmm, OK, then it wasn't an issue with the original implementation. Might be why it was missed in that testing.